File: templates/golang/persistent_subscription.go
// KurrentDB Go Client Example - Persistent Subscriptions
package main

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/google/uuid"
	"github.com/kurrent-io/KurrentDB-Client-Go/kurrentdb"
)

// OrderCreated represents an order creation event
type OrderCreated struct {
	OrderID string  `json:"orderId"`
	Amount  float64 `json:"amount"`
}

func main() {
	ctx := context.Background()

	// === CONNECTION ===
	settings, err := kurrentdb.ParseConnectionString("kurrentdb://admin:changeit@localhost:2113?tls=false")
	if err != nil {
		panic(err)
	}

	client, err := kurrentdb.NewClient(settings)
	if err != nil {
		panic(err)
	}
	defer client.Close()

	fmt.Println("Connected to KurrentDB")

	streamName := "orders"
	groupName := "order-processor"

	// === CREATE PERSISTENT SUBSCRIPTION ===
	err = client.CreatePersistentSubscription(
		ctx,
		streamName,
		groupName,
		kurrentdb.PersistentStreamSubscriptionOptions{},
	)
	if err != nil {
		// Check if already exists
		fmt.Printf("Persistent subscription '%s' may already exist: %v\n", groupName, err)
	} else {
		fmt.Printf("Created persistent subscription '%s' on stream '%s'\n", groupName, streamName)
	}

	// === APPEND SOME TEST EVENTS ===
	for i := 0; i < 3; i++ {
		order := OrderCreated{
			OrderID: uuid.New().String(),
			Amount:  10.00 * float64(i+1),
		}

		data, _ := json.Marshal(order)

		eventData := kurrentdb.EventData{
			EventID:     uuid.New(),
			ContentType: kurrentdb.ContentTypeJson,
			EventType:   "OrderCreated",
			Data:        data,
		}

		_, err := client.AppendToStream(
			ctx,
			streamName,
			kurrentdb.AppendToStreamOptions{},
			eventData,
		)
		if err != nil {
			panic(err)
		}
		fmt.Printf("Appended OrderCreated event #%d\n", i+1)
	}

	// === SUBSCRIBE TO PERSISTENT SUBSCRIPTION ===
	fmt.Println("\nSubscribing to persistent subscription...")

	subscription, err := client.SubscribeToPersistentSubscription(
		ctx,
		streamName,
		groupName,
		kurrentdb.SubscribeToPersistentSubscriptionOptions{},
	)
	if err != nil {
		panic(err)
	}
	defer subscription.Close()

	count := 0
	for {
		event := subscription.Recv()

		if event.EventAppeared != nil {
			fmt.Printf("  Processing: %s\n", event.EventAppeared.Event.EventType)
			fmt.Printf("  Data: %s\n", string(event.EventAppeared.Event.Data))

			// Simulate processing
			// ... your business logic here ...

			// Acknowledge success
			err := subscription.Ack(event.EventAppeared.Event)
			if err != nil {
				// Park failed event for later inspection
				subscription.Nack("Processing failed", kurrentdb.NackActionPark, event.EventAppeared.Event)
				fmt.Printf("  Parked event: %v\n", err)
			} else {
				fmt.Println("  Acknowledged event")
			}

			count++
			if count >= 3 {
				break
			}
		}

		if event.SubscriptionDropped != nil {
			fmt.Printf("Subscription dropped: %v\n", event.SubscriptionDropped.Error)
			break
		}
	}

	fmt.Println("\nDone!")
}

File: templates/rust/src/persistent_subscription.rs
//! KurrentDB Rust Client Example - Persistent Subscriptions

use kurrentdb::{Client, EventData, NakAction, PersistentSubscriptionOptions};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Serialize, Deserialize)]
struct OrderCreated {
    order_id: String,
    amount: f64,
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // === CONNECTION ===
    let settings = "kurrentdb://admin:changeit@localhost:2113?tls=false".parse()?;
    let client = Client::new(settings)?;

    println!("Connected to KurrentDB");

    let stream_name = "orders";
    let group_name = "order-processor";

    // === CREATE PERSISTENT SUBSCRIPTION ===
    match client
        .create_persistent_subscription(stream_name, group_name, &Default::default())
        .await
    {
        Ok(_) => {
            println!(
                "Created persistent subscription '{}' on stream '{}'",
                group_name, stream_name
            );
        }
        Err(e) => {
            println!(
                "Persistent subscription '{}' may already exist: {}",
                group_name, e
            );
        }
    }

    // === APPEND SOME TEST EVENTS ===
    for i in 0..3 {
        let order = OrderCreated {
            order_id: Uuid::new_v4().to_string(),
            amount: 10.0 * (i + 1) as f64,
        };

        let event = EventData::json("OrderCreated", &order)?.id(Uuid::new_v4());

        client
            .append_to_stream(stream_name, &Default::default(), event)
            .await?;

        println!("Appended OrderCreated event #{}", i + 1);
    }

    // === SUBSCRIBE TO PERSISTENT SUBSCRIPTION ===
    println!("\nSubscribing to persistent subscription...");

    let mut subscription = client
        .subscribe_to_persistent_subscription(stream_name, group_name, &Default::default())
        .await?;

    let mut count = 0;
    loop {
        match subscription.next().await {
            Ok(event) => {
                let original = event.get_original_event();
                println!("  Processing: {}", original.event_type);

                let data: OrderCreated = original.as_json()?;
                println!("  Data: {:?}", data);

                // Simulate processing
                // ... your business logic here ...

                // Acknowledge success
                match subscription.ack(&event).await {
                    Ok(_) => {
                        println!("  Acknowledged event");
                    }
                    Err(e) => {
                        // Park failed event for later inspection
                        subscription.nack(&event, NakAction::Park).await?;
                        println!("  Parked event: {}", e);
                    }
                }

                count += 1;
                if count >= 3 {
                    break;
                }
            }
            Err(e) => {
                println!("Subscription error: {}", e);
                break;
            }
        }
    }

    println!("\nDone!");
    Ok(())
}
